lib/itertools/stream/stream.go

```diff
@@ -0,0 +1,459 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package stream
+
+import (
+	"errors"
+	"io"
+	"iter"
+
+	"github.com/gravitational/trace"
+)
+
+// Stream is an alias for a fallible iterator. The Stream type alias and related
+// utilities follow a specific model that may not be appropriate for all types
+// that match the underlying signature. In particular, streams are modeled as the iterator
+// equivalent of a function of the form `func() ([]T, error)`. This means that streams
+// are expected to yield at most one error, and to not yield any additional values after
+// an error has been produced. The combinators in this package short-circuit on the
+// first error encountered, including those that handle multiple substreams.
+type Stream[T any] = iter.Seq2[T, error]
+
+// Func builds a stream from a closure. The supplied closure *must*
+// return io.EOF if no more items are available. Failure to return io.EOF
+// (or some other error) may cause infinite loops. Cleanup functions may
+// be optionally provided which will be run on close. If wrapping a
+// paginated API, consider using PageFunc instead.
+func Func[T any](fn func() (T, error)) Stream[T] {
+	return func(yield func(T, error) bool) {
+		for {
+			item, err := fn()
+			if err != nil {
+				if !errors.Is(err, io.EOF) {
+					var zero T
+					yield(zero, trace.Wrap(err))
+				}
+				return
+			}
+
+			if !yield(item, nil) {
+				return
+			}
+		}
+	}
+}
+
+// Collect aggregates a stream into a slice. If an error is hit, the
+// items observed thus far are still returned, but they may not represent
+// the complete set.
+func Collect[T any](stream Stream[T]) ([]T, error) {
+	var c []T
+	for item, err := range stream {
+		if err != nil {
+			return c, err
+		}
+		c = append(c, item)
+	}
+	return c, nil
+}
+
+// CollectPages aggregates a paginated stream into a slice. If an error
+// is hit, the pages observed thus far are still returned, but they may not
+// represent the complete set.
+func CollectPages[T any](stream Stream[[]T]) ([]T, error) {
+	var c []T
+	for page, err := range stream {
+		if err != nil {
+			return c, err
+		}
+		c = append(c, page...)
+	}
+	return c, nil
+}
+
+// FilterMap maps a stream of type A into a stream of type B, filtering out
+// items when fn returns false.
+func FilterMap[A, B any](stream Stream[A], fn func(A) (B, bool)) Stream[B] {
+	return func(yield func(B, error) bool) {
+		for item, err := range stream {
+			if err != nil {
+				var zero B
+				yield(zero, trace.Wrap(err))
```

Comment:
nit: normalize `*new(T)` as the universal zero value expression

```suggestion
				yield(*new(B), trace.Wrap(err))
```